### PR TITLE
k2_and_espnet2

### PR DESCRIPTION
**Pull request by glynpu**
As promised in #3176, I start to integrate k2 in espnet2.

Because I am not familar with espnet, perheps you could give me some suggestions to refactor current code before other modules are implemented. @sw005320 @takaaki-hori 
Currently, this pr is an implementation of k2 ctc decoding with the [SOTA espnet released model](https://zenodo.org/record/4604066#.YPUlxOgzZPY).

Plan of whole task:

 test-clean wer | encoder + k2-ctc-decode without rescore |encoder + k2-ctc-decode without rescore+3-gram+4-gram| encoder + k2 ctc decode +3-gram+4-gram+ 6-layer decoder nbest rescore | encoder + k2-ctc-decode +3-gram+4-gram+ transformer lm nbest rescore | encoder + ctc decode + 6-layer-decoder nbest rescore + transformer lm nbest rescore| encoder + k2 ctc decode +3-gram+4-gram+ 6-layer-decoder nbest rescore + transformer lm nbest rescore
-- | -- | -- | -- | --|--|--
baseline: current espnet SOTA result | * | * | * |*| 2.1% decoder algorithm in espnet|*
 decode with k2| 3.0 | to be implemented | to be implemented | to be implemented|*|  to be implemented

And the final goal of this task may combine encoder + transformer decoder + transformer lm + 3-gram + 4-gram together with k2. The whole process may be implementated in this single pr, module by module, or in seperated prs. 

Following is the detail result of current pr, which should be easily reproduced.

### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave/test_clean|2620|52576|97.4|2.4|0.2|0.3|3.0|34.7|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave/test_clean|2620|281530|99.3|0.3|0.3|0.4|1.0|34.7|

### CODE REVIEW

**Comment by sw005320** [comment]
This is very cool!
I’m in the vacation, but the other people will give you more suggestions.

**Comment by pengchengguo** [comment]
Hi @glynpu,

Thanks so much for your efforts.

Could we make `k2_decode.sh` as a template file for all ASR corpus, like `egs2/TEMPLATE/asr.sh`? Or we can integrate the k2_decode into the `asr.sh` as an optional stage. By doing so, we can remove both stage 0 and stage 2, which are already included in `egs2/TEMPLATE/asr.sh`.

File: espnet2/bin/k2_asr_inference.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+            batch = {k: v[0] for k, v in batch.items() if not k.endswith("_lengths")}
+
+            # 1-best list of (text, token, token_int)
+            results = speech2text(**batch)
+
+            # Only supporting batch_size==1
+            key = keys[0]
+            for text, token, token_int in results:
+
+                # Write the result to each file
```
Comment (ShigekiKarita) [inline espnet2/bin/k2_asr_inference.py:340]: Is it difficult to get score/confidence from K2?

File: espnet2/bin/k2_asr_inference.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+                                                                   device=best_paths.device))
+    return k2r.to_list(aux_labels)
+
+
+def invert_permutation(indices: torch.Tensor) -> torch.Tensor:
+    ans = torch.zeros(indices.shape, device=indices.device, dtype=torch.long)
+    ans[indices] = torch.arange(0, indices.shape[0], device=indices.device)
+    return ans
+
+class Speech2Text:
```
Comment (ShigekiKarita) [inline espnet2/bin/k2_asr_inference.py:111]: Can you create unittests like existing S2T? https://github.com/espnet/espnet/blob/e60d1786879839234cf3db8aad1f71af6f93333d/test/espnet2/bin/test_asr_inference.py#L107

**Comment by ShigekiKarita** [comment]
This is concise and cool! After adding the TEMPLATE, and unittests, and fixing CI errors (try `$ black espnet espnet2 test utils setup.py egs*/*/*/local egs2/TEMPLATE/asr1/pyscripts`), I think we can merge this PR.

File: espnet2/bin/k2_asr_inference.py (hunk 1 above)
Comment (csukuangfj) [reply espnet2/bin/k2_asr_inference.py:340]: No, it is very easy.

https://github.com/espnet/espnet/blob/aadf45baeb26a4900c4ac6be0f286a51ff57de64/espnet2/bin/k2_asr_inference.py#L225

```python
scores = best_paths.get_tot_scores(use_double_scores=True, log_semiring=False)
```
returns the scores for the best paths in the batch.

Comment (ShigekiKarita) [reply espnet2/bin/k2_asr_inference.py:340]: Cool! Let's save the value like this
https://github.com/espnet/espnet/blob/e60d1786879839234cf3db8aad1f71af6f93333d/espnet2/bin/asr_inference.py#L338

**Comment by glynpu** [comment]
> Could we make `k2_decode.sh` as a template file for all ASR corpus, like `egs2/TEMPLATE/asr.sh`? Or we can integrate the k2_decode into the `asr.sh` as an optional stage. By doing so, we can remove both stage 0 and stage 2, which are already included in `egs2/TEMPLATE/asr.sh`.

fixed by adding  a config use_k2=False in egs2/TEMPLATE/asr.sh. 

Now an example command to use k2 ctc_decoder is:
```
./run.sh \
  --skip_data_prep true \
  --skip_train true \
  --download_model kamo-naoyuki/librispeech_asr_train_asr_conformer6_n_fft512_hop_length256_raw_en_bpe5000_scheduler_confwarmup_steps40000_optim_conflr0.0025_sp_valid.acc.ave \
  --use_k2 true \
  --gpu_inference true
```

File: espnet2/bin/k2_asr_inference.py (hunk 1 above)
Comment (glynpu) [reply espnet2/bin/k2_asr_inference.py:366]: fixed.

**Comment by glynpu** [comment]
> This is concise and cool! After adding the TEMPLATE, and unittests, and fixing CI errors (try `$ black espnet espnet2 test utils setup.py egs*/*/*/local egs2/TEMPLATE/asr1/pyscripts`), I think we can merge this PR.

A unitest for class `k2Speech2Text` is added. And fixed some CI errors.
The remaining several CI errors is caused by  no-compatible k2 for some system or pytorch configs.

**Comment by sw005320** [comment]
> > This is concise and cool! After adding the TEMPLATE, and unittests, and fixing CI errors (try `$ black espnet espnet2 test utils setup.py egs*/*/*/local egs2/TEMPLATE/asr1/pyscripts`), I think we can merge this PR.
> 
> A unitest for class `k2Speech2Text` is added. And fixed some CI errors.
> The remaining several CI errors is caused by no-compatible k2 for some systems or pytorch configs.

@ShigekiKarita, do you have any suggestions?
Should we make an installation script for k2 like fariseq? (e.g., https://github.com/espnet/espnet/blob/master/tools/installers/install_fairseq.sh)
Then, we can skip the unit test by using `is_torch_1_5_1_plus = LooseVersion(torch.__version__) >= LooseVersion("1.5.1")` like https://github.com/espnet/espnet/blob/6e131379ffae210ef76eae3cb768e3ebc9b3f9bf/test/espnet2/enh/test_espnet_model.py#L196-L197 ?

**Comment by csukuangfj** [comment]
> The remaining several CI errors is caused by no-compatible k2 for some systems or pytorch configs.

That should be fixed by https://github.com/k2-fsa/k2/pull/792

**Comment by sw005320** [comment]
OK, thanks!
If it is ready, please let me know.
I'll rerun the CI test, then.

**Comment by glynpu** [comment]
>OK, thanks!
If it is ready, please let me know.
I'll rerun the CI test, then.

All ci errors are fixed following the way you suggested @sw005320 . 
Now fixing codecov related errors.

**Comment by glynpu** [comment]
Could you give me some suggestions to fix codecov related errors? @sw005320 @ShigekiKarita @pengchengguo 

Questions 1 about codecov/project: 80.84(-0.04),
Why this pr results in [a decrease of espnet2/bin/asr_inference.py](https://app.codecov.io/gh/espnet/espnet/compare/3358/changes)? My modifications should have no relationship with asr_inference.py.

Questions 2, is there a good way to fix  (codecov/patch Failing after 1s — 77.83% of diff hit (target 80.87%))? Could you give me an example to follow.

**Comment by sw005320** [comment]
Great!
You do not have to care codecov.

Did you also add tools/installers/install_k2.sh?
I cannot find it.

**Comment by glynpu** [comment]
>Did you also add tools/installers/install_k2.sh?
I cannot find it.

Add it in the latest commit. Sorry for fogetting that.

**Comment by pengchengguo** [comment]
Thanks @glynpu, it is good to me now!

**Comment by sw005320** [comment]
@glynpu, thanks!
Is this PR ready?
Can you remove WIP in the issue title?
I’ll merge it then.

**Comment by glynpu** [comment]
>Is this PR ready?
Can you remove WIP in the issue title?
I’ll merge it then.

Yes, if we all agree. @sw005320 @ShigekiKarita 

I will implemented lattice and transformer-decoder rescoring parts in other following prs.

**Comment by sw005320** [comment]
Thanks @glynpu!
I just merged this PR.

This would be a very important function for espnet.
I’m very excited about this work.
I may ask you to also touch the main README.md to advertise this function, later.

@ShigekiKarita, we welcome your further comments. You can do in in this PR or follow-up PRs.

**Comment by ShigekiKarita** [comment]
LGTM. Thanks for your hard work.

> I will implemented lattice and transformer-decoder rescoring parts in other following prs.

I'm also looking forward your future PRs!

**Comment by sw005320** [comment]
Hi @glynpu,

It seems that k2 has some issues in https://github.com/espnet/espnet/runs/3258206278?check_suite_focus=true#step:8:35 (it used to be working two days ago).
Do you know what happens?

**Comment by csukuangfj** [comment]
@sw005320 

From the log of GitHub actions
https://github.com/espnet/espnet/runs/3258206278?check_suite_focus=true#step:6:3640

```
## Package Plan ##

  environment location: /__w/espnet/espnet/tools/venv/envs/espnet

  added / updated specs:
    - cpuonly
    - k2
    - pytorch=1.9.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    cudatoolkit-10.2.89        |       hfd86e86_1       365.1 MB
    k2-1.3.dev20210805         |cuda10.2_py3.6_torch1.9.0        59.9 MB  k2-fsa
    ------------------------------------------------------------
                                           Total:       424.9 MB

The following NEW packages will be INSTALLED:

  cudatoolkit        pkgs/main/linux-64::cudatoolkit-10.2.89-hfd86e86_1
  k2                 k2-fsa/linux-64::k2-1.3.dev20210805-cuda10.2_py3.6_torch1.9.0
```

https://github.com/espnet/espnet/runs/3258206278?check_suite_focus=true#step:8:48

```
espnet2/bin/k2_asr_inference.py:11: in <module>
    import k2
tools/venv/envs/espnet/lib/python3.6/site-packages/k2/__init__.py:2: in <module>
    from _k2 import DeterminizeWeightPushingType
E   ImportError: libtorch_cuda.so: cannot open shared object file: No such file or directory
========
```

It seems that conda installed a version of k2 compiled with CUDA. I am not sure why conda does not pick the CPU version.
I guess the reason is that the CUDA version in https://anaconda.org/k2-fsa/k2/files is newer than the CPU version.

I am building the CPU version via https://github.com/k2-fsa/k2/runs/3263625412?check_suite_focus=true
Once the building process completes (it takes about 1 hour and 10 minutes), @sw005320  could you restart the failed GitHub action? I think the error will disappear.
If not, I will try to find another alternative to fix it.

**Comment by sw005320** [comment]
OK!

**Comment by csukuangfj** [comment]
@sw005320 
Could you restart the failed GitHub action? conda contains the latest CPU version of k2 now.

**Comment by sw005320** [comment]
The issue is fixed!
Thanks!

**Comment by glynpu** [comment]
> could you share your plan for the rest of the items?
> no pressure, but I just would like to know the plan.

Sure. 
A rough plan contains following steps:
0. encoder + ctc decocer （where we are）
1. + support batch decoding
2. + support ctc_topo + 3-gram decoding, which will be referred as TLG decoding and T is short for "Token".
3. + 4-gram lattice rescore 
4. + ransformer-decoder nbest rescore
5. + neural net language model nbest rescore

So, the whole decoding pipeline is (encoder + TLG decoding)  and other theree rescore parts: one is 4-gram lattice rescore and other two are nbest rescore.

Hopefully, all previously mentioned parts will be implemented before next Sunday, Aug. 22. (Make a promiss here to keep myself more active in this task. I was occupied by other works in the past two weeks. )

**Comment by sw005320** [comment]
Thanks!
Actually, many people (including me!) are excited about this PR, and are looking forward to the follow up PRs.
I’m also thinking of making it work in a  (segment-wise) streaming fashion.

 > support ctc_topo + 3-gram decoding, which will be referred as TLG decoding and T is short for "Token".

Do you mean that 3-gram LM is based on a word unit?

**Pull request merged**